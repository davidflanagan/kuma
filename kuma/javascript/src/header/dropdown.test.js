//@flow
import React from 'react';
import { act, create } from 'react-test-renderer';
import Dropdown from './dropdown.jsx';

jest.useFakeTimers();

test('Dropdown closed and open snapshots', () => {
    const dropdown = create(
        <Dropdown label="Test">
            <li>foo</li>
            <li>bar</li>
        </Dropdown>
    );

    let closedTree = dropdown.toJSON();
    let closedString = JSON.stringify(closedTree);

    // Expect the menu to be closed and take a snapshot.
    // Even though the menu is closed, we expect the content to
    // be in the document so web crawlers and screen readers see it
    expect(closedString).toContain('▼');
    expect(closedString).toContain('foo');
    expect(closedString).toContain('bar');
    expect(closedTree).toMatchSnapshot();

    // Fake a click on the menu label
    act(() => {
        dropdown.toJSON().children[0].props.onClick();
    });

    let openTree = dropdown.toJSON();
    let openString = JSON.stringify(openTree);

    // Now expect the menu to be open, and snapshot it in that state
    expect(openString).toContain('▲');
    expect(closedString).toContain('foo');
    expect(closedString).toContain('bar');
    expect(openTree).toMatchSnapshot();

    // Fake a click on the document body to dismiss the menu
    act(() => {
        document.body && document.body.dispatchEvent(new MouseEvent('click'));
        jest.runAllTimers();
    });

    // Expect the menu to be closed, checking deep equality
    expect(JSON.stringify(dropdown.toJSON())).toEqual(closedString);

    // Open the menu again
    act(() => {
        dropdown.toJSON().children[0].props.onClick();
    });

    // Verify that it is opened
    expect(JSON.stringify(dropdown.toJSON())).toEqual(openString);

    // Fake a keyboard event
    act(() => {
        document.body &&
            document.body.dispatchEvent(
                new KeyboardEvent('keydown', { key: 'A' })
            );
        jest.runAllTimers();
    });
    // The menu should still be open
    expect(JSON.stringify(dropdown.toJSON())).toEqual(openString);

    // Finally, fake the escape key
    act(() => {
        document.body &&
            document.body.dispatchEvent(
                new KeyboardEvent('keydown', { key: 'Escape' })
            );
        jest.runAllTimers();
    });
    // And expect the menu to have closed
    expect(JSON.stringify(dropdown.toJSON())).toEqual(closedString);
});
