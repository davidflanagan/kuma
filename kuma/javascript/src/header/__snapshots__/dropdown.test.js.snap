// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown closed and open snapshots 1`] = `
.emotion-6 {
  position: relative;
  pointer-events: auto;
  margin-left: -5px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  white-space: nowrap;
  padding: 0 5px;
  border: none;
}

.emotion-2:hover {
  background-color: #eee;
}

.emotion-2:focus {
  outline-offset: -3px;
}

.emotion-0 {
  font-size: 75%;
  padding-left: 2px;
}

.emotion-4 {
  position: absolute;
  z-index: 100;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  background-color: white;
  border: solid #83d0f2 1.5px;
  box-shadow: 3px 3px 5px rgba(0,0,0,0.25);
  padding: 10px;
  min-width: 100%;
  display: none;
}

.emotion-4 li {
  padding: 5px;
  white-space: nowrap;
}

<div
  className="emotion-6 emotion-7"
>
  <button
    className="emotion-2 emotion-3"
    onClick={[Function]}
  >
    Test
    <span
      className="emotion-0 emotion-1"
    >
      ▼
    </span>
  </button>
  <ul
    className="dropdown-menu emotion-4 emotion-5"
  >
    <li>
      foo
    </li>
    <li>
      bar
    </li>
  </ul>
</div>
`;

exports[`Dropdown closed and open snapshots 2`] = `
.emotion-6 {
  position: relative;
  pointer-events: auto;
  margin-left: -5px;
}

.emotion-0 {
  font-size: 75%;
  padding-left: 2px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  white-space: nowrap;
  padding: 0 5px;
  border: none;
  background-color: #83d0f2;
}

.emotion-2:hover {
  background-color: #eee;
}

.emotion-2:focus {
  outline-offset: -3px;
}

.emotion-2:hover {
  background-color: #83d0f2;
}

.emotion-4 {
  position: absolute;
  z-index: 100;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  background-color: white;
  border: solid #83d0f2 1.5px;
  box-shadow: 3px 3px 5px rgba(0,0,0,0.25);
  padding: 10px;
  min-width: 100%;
}

.emotion-4 li {
  padding: 5px;
  white-space: nowrap;
}

<div
  className="emotion-6 emotion-7"
>
  <button
    className="emotion-2 emotion-3"
    onClick={null}
  >
    Test
    <span
      className="emotion-0 emotion-1"
    >
      ▲
    </span>
  </button>
  <ul
    className="dropdown-menu emotion-4 emotion-5"
  >
    <li>
      foo
    </li>
    <li>
      bar
    </li>
  </ul>
</div>
`;
